library/ccm.c

```diff
@@ -253,62 +211,290 @@ static int ccm_auth_crypt( mbedtls_ccm_context *ctx, int mode, size_t length,
      * 7 .. 3   0
      * 2 .. 0   q - 1
      */
-    ctr[0] = q - 1;
-    memcpy( ctr + 1, iv, iv_len );
-    memset( ctr + 1 + iv_len, 0, q );
-    ctr[15] = 1;
+    memset( ctx->ctr, 0, 16);
+    ctx->ctr[0] = ctx->q - 1;
+    memcpy( ctx->ctr + 1, iv, iv_len );
+    memset( ctx->ctr + 1 + iv_len, 0, ctx->q );
+    ctx->ctr[15] = 1;
 
     /*
-     * Authenticate and {en,de}crypt the message.
+     * See ccm_calculate_first_block_if_ready() for block layout description
+     */
+    memcpy( ctx->y + 1, iv, iv_len );
+
+    ctx->state |= CCM_STATE__STARTED;
+    return ccm_calculate_first_block_if_ready(ctx);
+}
+
+int mbedtls_ccm_set_lengths( mbedtls_ccm_context *ctx,
+                             size_t total_ad_len,
+                             size_t plaintext_len,
+                             size_t tag_len )
+{
+    /*
+     * Check length requirements: SP800-38C A.1
+     * Additional requirement: a < 2^16 - 2^8 to simplify the code.
+     * 'length' checked later (when writing it to the first block)
      *
-     * The only difference between encryption and decryption is
-     * the respective order of authentication and {en,de}cryption.
+     * Also, loosen the requirements to enable support for CCM* (IEEE 802.15.4).
      */
-    len_left = length;
-    src = input;
-    dst = output;
+    if( tag_len == 2 || tag_len > 16 || tag_len % 2 != 0 )
+        return( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    if( total_ad_len >= 0xFF00 )
+        return( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    ctx->plaintext_len = plaintext_len;
+    ctx->add_len = total_ad_len;
+    ctx->tag_len = tag_len;
+    ctx->processed = 0;
+
+    ctx->state |= CCM_STATE__LENGHTS_SET;
+    return ccm_calculate_first_block_if_ready(ctx);
+}
 
-    while( len_left > 0 )
+int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
+                           const unsigned char *add,
+                           size_t add_len )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char i;
+    size_t olen, use_len, offset;
+
+    if( ctx->state & CCM_STATE__ERROR )
+    {
+        return ret;
+    }
+
+    if( ctx->add_len > 0 && add_len > 0 )
```

Comment:
(Continued from https://github.com/ARMmbed/mbedtls/pull/4708#discussion_r684254230)

If `ctx->add_len == 0` and `add_len == 0`, i.e. if empty AD is expected but the application passes non-empty AD, this function should return `MBEDTLS_ERR_CCM_BAD_INPUT`, but this isn't happening.

I think removing `ctx->add_len > 0 &&` would fix that.

Please add a non-regression test.